File: TUSHAR-JAVA/BinarySearch.java
//! wap a binary search algorithm in java.



public class BinarySearch {
    // Performs iterative binary search on sorted array a for target t
    static int bs(int[] a, int t) {
        int l = 0;                 // left boundary (inclusive)
        int r = a.length - 1;      // right boundary (inclusive)
        while (l <= r) {           // continue while search space is not empty
            int m = (l + r) >>> 1; // mid index (unsigned shift avoids overflow)
            if (a[m] == t)         // if middle element is target
                return m;          // return its index
            if (a[m] < t)          // if middle element less than target
                l = m + 1;         // discard left half including m
            else
                r = m - 1;         // discard right half including m
        }
        return -1;                 // target not found
    }

    public static void main(String[] args) {
        int[] a = {1,2,3,4,5,6,7,8,9,10}; // sorted input array
        System.out.println(bs(a, 5));     // search for 5, expect index 4
    }
}




// ? output
// ? 







//?     Time complexity: Best case: 
// ?    Best case: O(1) (found at first check)
//?    Average case: O(log n)
//?    Worst case: O(log n)


//?    Space complexity: O(1)
//?    Iterative: O(1) (no extra space)
//?    Recursive: O(log n) (due to call stack)

File: BAIBHAB-PYTHON/binary_search.py
# binary search algorithm: (ARRAY HAS TO BE SORTED!)
# 1. divide the array into two halves by finding the middle index
# 2. compare mid elememt with the key
# 3. if key == mid element
# 4.      return index
# 5. else if key < mid element
# 6.      search from arr[0] to arr[mid - 1] -> left half
# 7. else if key > mid element
# 8.      search from arr[mid + 1] to arr[n-1] -> right half
# 9. else
# 10.     return -1 -> element not found

# Time: O(logN)
# Space: O(1)

def binary_search(arr: list[int], key: int) -> int:
    low, high = 0, len(arr) - 1

    while low <= high:
        # mid = (low+high) // 2  # use floor division to get the mid index
        # prevents integer overflow for huge array inputs
        mid = low + ((high - low) // 2)

        if arr[mid] > key:
            high = mid - 1
        elif arr[mid] < key:
            low = mid + 1
        else:
            return mid  # element found at that index
    return -1  # element not found


def main():
    arr = [2, 3, 4, 10, 40]
    key = 10

    # function call
    result = binary_search(arr, key)
    # print result
    if result != -1:
        print(f"Element found at index {result}")
    else:
        print("Element not found!")


if __name__ == "__main__":
    main()
